includes/class-sensei-course.php

```diff
@@ -681,9 +683,47 @@ public function add_column_headings( $defaults ) {
 			$new_columns['date'] = $defaults['date'];
 		}
 
+		foreach ( $defaults as $column_key => $column_value ) {
+			if ( ! isset( $new_columns[ $column_key ] ) ) {
+				$new_columns[ $column_key ] = $column_value;
+			}
+		}
+
 		return $new_columns;
 	} // End add_column_headings()
 
+	/**
+	 * Hide all columns by default, leaving only a default list.
+	 *
+	 * @access public
+	 * @since  3.1.0-dev
+	 * @param  array     $hidden_columns
+	 * @param  WP_Screen $screen
+	 * @return array     $hidden_columns
+	 */
+	public function set_default_visible_columns( $hidden_columns, $screen ) {
```

Comment:
It's not clear to me what this function does and how to test it. Could you clarify?

Reply:
In case the purpose of this function is not clear yet at this point. It is a callback for the 'default_hidden_columns' filter hook and sets all columns to be hidden by default except the ones sensei LMS want's shown by default.

In that way, it would setup only the list columns to be visible that would have been shown before this pull request was added.  So any columns not belonging to sensei lms (the defaults) is set to be hidden by default.

This function won't affect any screen options a user has set herself, since the filter is only called until a user changes the screen options (the admin-pulldown where the user can select what columns to hide or show). It just sets a default presentation after add_column_headings() has added the specific sensei-columns and determined the order of the columns.

In the database the settings a user sets are in the user-meta "manage{screen-id}columnshidden". When that entry is present (and not empty) then this function/method is no longer called.

Reply:
I tested this, but either it's not working as intended, or I'm not sure we need it. I enabled the Polylang plugin on Sensei LMS 3.6 and had the _Lessons_ set like so:

![Screen Shot 2020-12-14 at 10 58 42 AM](https://user-images.githubusercontent.com/1190420/102103735-5958a780-3dfb-11eb-8db8-79ce85633847.jpg)

After updating to the code in this PR, I lost the language flags on the _Lessons_ and _Questions_ pages, because the languages aren't selected anymore in _Screen Options_. That's not ideal, as people will wonder where their columns went.

Also, by doing this we won't get to see the _Modules_ column on the _Courses_ and _Lessons_ page, which is actually a beneficial one to have. _Comments_ are also shown on _Lessons_, which I'm OK with.

So, all this to say that I think we should probably remove these hook handlers.

Reply:
Hi @donna. Sorry for the mistake. The set_default_visible_columns method bases what columns to leave visible on an internal array. I had forgotten to add the columns you where missing to that array. That should be fixed in the latest update of this fix. The default visible columns could be made a setting or property to make it more obvious and manipulable. Then when f.i. sensei content drip will add the drip schedule column to be show by default, it is able to do so.

I have looked at polylang in a glance, but it will probably will be a timing issue. No columns are deleted, they are at most hidden by default.  Let me look at it first. 

The problem of not having the set_default_visible_columns method is potentially having a lot more columns visible by default or entirely removing columns like the previous version of add_column_headings did.

I will come back to this soon.

Reply:
@donnapep I've updated the fix so it will keep the polylang language flag columns visible too. They are all called 'language_' with a country code suffix.

Reply:
@hansschuijff Thanks for the changes. I think hardcoding to look for `language_` is a fairly brittle solution. While this works for Polylang, what if there are other plugins out there that also work a similar way? As such, I'm not sure this is optimal.

> The problem of not having the set_default_visible_columns method is potentially having a lot more columns visible by default or entirely removing columns like the previous version of add_column_headings did.

It seems to me that we have one of 2 choices:
1. Potentially show too many columns.
2. Potentially hide columns that used to show before.

Of these 2 options, #1 seems much lower risk than #2, given that we'd only be adding more columns and not taking columns away people could be used to seeing.

Given this, and the brittleness of the Polylang fix, I'd still advocate for removing the `set_default_visible_columns` functions.

Reply:
Also, I just tested with Yoast with the `default_hidden_columns` filters disabled and was not seeing any issues with columns being removed.

Reply:
@donnapep I agree that it can be a slippery slope when starting to accommodate for other plugins, so in the end it's a decision and I'm fine either way.

The current production version bluntly removes columns of other plugins instead of hiding them, so there the user needs a developer to retrieve columns. In this version it is just a question of showing or hiding other columns. All columns stay present and are ordered so that the preferred columns are in front. That's already an improvement.

But as it is your plugin, the choice comes mostly down to what you think will be best for your users. 

Other plugins can still unhide their columns if they wish (same hook, higher prio), but users may not know how to retrieve hidden columns (although admin users will probably easily find guidance in that too).

You're right that making special accommodations for other plugins is always a bit tricky, since others may want that too. But it wont be the first time a plugin does that for a popular counterpart. It's just a decison.

1. leave it to other plugins to unhide columns
2. selectively hide columns
3. don't hide any columns

Option 3 is the option of least maintenance, and probably more dummy proof for users. So I will support that, if the decision is made. Just let me know, then we can finish this pr.

Reply:
@donnapep the issue concerning Yoast WP-Seo has been solved in version 15.4 of the Yoast plugin. So no problem there anymore.